### PR TITLE
Debugging techniques document with a first practice described

### PR DESCRIPTION
**Pull request by nikita-leonov**
Added Debugging Techniques document with a description of debugging compiler type errors referenced in an issue #2381

### CODE REVIEW

File: Documentation/DebuggingTechniques.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    .on(next: { answer in
+        return _
+    })
+    .startWithCompleted {
+        print("Completed.")
+    }
+```
+
+The code above will not compile with a following error on a `print` call `error: ambiguous reference to member 'print'
+print("Completed.")` To find an actual source of a error signals chain need to be broken down apart witha an explicit definition of clojure types on each of the steps:
```
Comment (saniul) [inline Documentation/DebuggingTechniques.md]: s/witha an/with an

File: Documentation/DebuggingTechniques.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+This document contains list of debugging techniques and infrastructure that helpful for debugging Reactive Cocoa applications.
+
+#### Unscrambling Swift compiler errors
+
+Type inferrence could be a source of hard-to-debug compiler errors. There are two potential places to be wrong when type inferrence used:
+
+1. Definition of type inferred variable
+2. Consumption of type inferred variable
+
+In both cases erros are related to incorret assumptions about type. Such issues are common for a ReactiveCocoa applications as it is all about operations over data and related types. With a current state of a Swift compiler type related errors could be really misleading, especially when error happens in the middle of a signal chain.
```
Comment (saniul) [inline Documentation/DebuggingTechniques.md]: s/incorret/incorrect

File: Documentation/DebuggingTechniques.md (hunk 1 above)
Comment (saniul) [inline Documentation/DebuggingTechniques.md]: s/clojure/closure

File: Documentation/DebuggingTechniques.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+let initialProducer = SignalProducer<Int, NoError>.init(value:42)
+let sideEffectProducer = initialProducer.on(next: { (answer: Int) in
+    return _
+})
+let disposable = sideEffectProducer.startWithCompleted {
+    print("Completed.")
+}
+```
+
+The code above will not compile too, but with a error `error: cannot convert value of type '(_) -> _' to expected argument type '(Int -> ())?'` on definition of `on` clojure. This gives enough of information to locate unexpected `return _` since `on` clojure should not have any return value.
```
Comment (saniul) [inline Documentation/DebuggingTechniques.md]: s/clojure/closure

**Comment by nikita-leonov** [comment]
Thanks @saniul :) @neilpa planned to provide service of proofreading, but you did this earlier. Thanks for your help.

File: Documentation/DebuggingTechniques.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+This document contains list of debugging techniques and infrastructure that helpful for debugging Reactive Cocoa applications.
+
+#### Unscrambling Swift compiler errors
+
+Type inferrence could be a source of hard-to-debug compiler errors. There are two potential places to be wrong when type inferrence used:
+
+1. Definition of type inferred variable
+2. Consumption of type inferred variable
+
+In both cases erros are related to incorrect assumptions about type. Such issues are common for a ReactiveCocoa applications as it is all about operations over data and related types. With a current state of a Swift compiler type related errors could be really misleading, especially when error happens in the middle of a signal chain.
```
Comment (iv-mexx) [inline Documentation/DebuggingTechniques.md]: *errors

File: Documentation/DebuggingTechniques.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Debugging Techniques
+
+This document contains list of debugging techniques and infrastructure that helpful for debugging Reactive Cocoa applications.
```
Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: We could shorten this to

> This document lists debugging techniques and infrastructure helpful for debugging ReactiveCocoa applications.

File: Documentation/DebuggingTechniques.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+#### Unscrambling Swift compiler errors
+
+Type inferrence could be a source of hard-to-debug compiler errors. There are two potential places to be wrong when type inferrence used:
```
Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: could by -> can be

File: Documentation/DebuggingTechniques.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+1. Definition of type inferred variable
+2. Consumption of type inferred variable
+
+In both cases errors are related to incorrect assumptions about type. Such issues are common for a ReactiveCocoa applications as it is all about operations over data and related types. With a current state of a Swift compiler type related errors could be really misleading, especially when error happens in the middle of a signal chain.
```
Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: Such issues are common for ~~a~~ ReactiveCocoa applications

Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: ~~With a current state of a Swift compiler type related errors could be really misleading,~~
The current state of the Swift compiler can cause misleading type errors,

File: Documentation/DebuggingTechniques.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+SignalProducer<Int, NoError>(value:42)
+    .on(next: { answer in
+        return _
+    })
+    .startWithCompleted {
+        print("Completed.")
+    }
+```
+
+The code above will not compile with a following error on a `print` call `error: ambiguous reference to member 'print'
```
Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: with ~~a~~ following
with the following

File: Documentation/DebuggingTechniques.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+print("Completed.")` To find an actual source of a error signals chain need to be broken down apart with an explicit definition of closure types on each of the steps:
```
Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: To find ~~an~~ **the** actual source of ~~a~~ error**s**, signal~~s~~ chain**s** need to be broken ~~down~~ apart. **Add** ~~with an~~ explicit definition**s** of closure types on each of the steps:

File: Documentation/DebuggingTechniques.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+let initialProducer = SignalProducer<Int, NoError>.init(value:42)
+let sideEffectProducer = initialProducer.on(next: { (answer: Int) in
+    return _
+})
+let disposable = sideEffectProducer.startWithCompleted {
+    print("Completed.")
+}
+```
+
+The code above will not compile too, but with a error `error: cannot convert value of type '(_) -> _' to expected argument type '(Int -> ())?'` on definition of `on` closure. This gives enough of information to locate unexpected `return _` since `on` closure should not have any return value.
```
Comment (neilpa) [inline Documentation/DebuggingTechniques.md]: with ~~a~~ **the** error

**Comment by neilpa** [comment]
Thanks a lot for tackling this! This is good stuff and I know it's not the easiest for non-native English speakers.

I'll merge this once those edits are in.

**Comment by nikita-leonov** [comment]
@neilpa corrections are done. Thanks for review :)

**Comment by neilpa** [comment]
:sparkler:

**Comment by nikita-leonov** [comment]
Amaaazing :) Thanks! Lets keep going, I kind of enjoy the velocity of how everything changing into the proper direction.
